### PR TITLE
Open a folder on android using intent

### PR DESCRIPTION
**Pull request by Dragicafit**
Undo https://github.com/ppy/osu/pull/4353 (Open osu! Folder only)

Proof : https://gyazo.com/11d1c6aff9fe8952e683fb237c5902dd
It doesn't open a file.
It opens a folder.

Based on https://github.com/xamarin/Essentials/pull/773/files#diff-e493aa2940119a81025b78a11720b7b9

This is the only way that works with [Cx file explorer](https://play.google.com/store/apps/details?id=com.cxinventor.file.explorer), [Solid Explorer](https://play.google.com/store/apps/details?id=pl.solidexplorer2), [X-plore](https://play.google.com/store/apps/details?id=com.lonelycatgames.Xplore):
```
action=ActionView
data=/path/to/directory/ 
mime=resource/folder 
```

action:
`ActionGetContent` opens the android file explorer but does nothing else
`ActionView` works (https://github.com/syncthing/syncthing-android/issues/838#issuecomment-275370227)

data:
`file://` is not supported since api 24 (https://stackoverflow.com/questions/38200282/android-os-fileuriexposedexception-file-storage-emulated-0-test-txt-exposed)
`content://` is not supported by any file explorer
`URI.Parse(filename)` works

mime:
`*/*` shows all the app but none will works
`text/csv` shows all the text editors but none will works
`resource/folder` is not official but is used by some file explorer to open a folder (https://github.com/syncthing/syncthing-android/issues/838)

extra:
[`org.openintents.extra.ABSOLUTE_PATH`](http://www.openintents.org/action/android-intent-action-view/file-directory) is not supported by any file eplorer

apps used: 
file explorer

SOURCE:
[similar issue](https://github.com/syncthing/syncthing-android/issues/838)
[wrong way to implement](https://stackoverflow.com/questions/17165972/android-how-to-open-a-specific-folder-via-intent-and-show-its-content-in-a-file)
[exemple of support in Amaze File Manager](https://github.com/TeamAmaze/AmazeFileManager/issues/419)
[`content://`](https://stackoverflow.com/questions/38200282/android-os-fileuriexposedexception-file-storage-emulated-0-test-txt-exposed)
[`var contentUri = Platform.GetShareableFileUri(request.File.FullPath);`](https://github.com/xamarin/Essentials/blob/master/Xamarin.Essentials/Platform/Platform.android.cs#L66)

### CODE REVIEW

File: osu.Framework.Android/AndroidGameHost.cs [hunk 1]
```diff
@@ -45,7 +50,30 @@ protected override Storage GetStorage(string baseName)
             => new AndroidStorage(baseName, this);
 
         public override void OpenFileExternally(string filename)
-            => throw new NotImplementedException();
+        {
+            if (!File.Exists(filename))
```
Comment (peppy) [inline osu.Framework.Android/AndroidGameHost.cs]: should this not still throw if a file is attempted to be opened?

Comment (Dragicafit) [reply osu.Framework.Android/AndroidGameHost.cs]: my bad, should I do something like that ?
```
if(File.Exists(filename)) 
{
    throw new NotImplementedException();
}               
else if(Directory.Exists(filename)) 
{
    openFolderExternally(filename);
}
else 
{
    ??
} 
```

Comment (peppy) [reply osu.Framework.Android/AndroidGameHost.cs]:
```csharp
if (directory exists)
{
   // do think
}

throw new NotImplementedException();
```

File: osu.Framework.Android/AndroidGameHost.cs [hunk 2]
```diff
@@ -45,7 +50,31 @@ protected override Storage GetStorage(string baseName)
...
 
         public override void OpenFileExternally(string filename)
-            => throw new NotImplementedException();
+        {
+            if (Directory.Exists(filename))
+            {
+                openFolderExternally(filename);
+                return;
+            }
+            throw new NotImplementedException();
```
Comment (swoolcock) [inline osu.Framework.Android/AndroidGameHost.cs]: This needs to be a nicer solution.  We can't have an app outright crashing if the path doesn't exist.

File: osu.Framework.Android/AndroidGameHost.cs [hunk 3]
```diff
@@ -45,7 +50,31 @@ protected override Storage GetStorage(string baseName)
...
+        }
+
+        private void openFolderExternally(string filename)
```
Comment (swoolcock) [inline osu.Framework.Android/AndroidGameHost.cs]: Is there a better way you can solve this that supports more than just directories?  The public method name is Open*File*Externally.

Comment (Dragicafit) [reply osu.Framework.Android/AndroidGameHost.cs]: OpenFileExternally is already supported in Xamarin.Essentials, so we just have to import the nuget (https://docs.microsoft.com/en-us/xamarin/essentials/launcher?tabs=android).
Or we can copy the source code from if you don't want to import it (https://github.com/xamarin/Essentials/blob/master/Xamarin.Essentials/Launcher/Launcher.android.cs#L33)

File: osu.Framework.Android/AndroidGameHost.cs (hunk 2 above)
Comment (Dragicafit) [reply osu.Framework.Android/AndroidGameHost.cs]: This has to be changed when openFileExternally will be created.

File: osu.Framework.Android/AndroidGameHost.cs (hunk 3 above)
Comment (peppy) [reply osu.Framework.Android/AndroidGameHost.cs]: Can you implementing via nuget? That sounds a lot nicer if it works.

**Comment by dragonGod97677** [comment]
How about MiXplorer? wanna test it but lazer cant run on my phone, it can be downloaded on  xda ( https://labs.xda-developers.com/store/app/com.mixplorer )

**Comment by peppy** [comment]
Specific file manager apps are out of scope of this.

**Comment by Dragicafit** [comment]
Now, if you install an app that can read a JSON file ([7Zipper](https://play.google.com/store/apps/details?id=org.joa.zipperplus7)), you can read the exported file in the editor.

**Comment by smoogipoo** [comment]
It appears like this has since been superseded.

**Pull request closed without merging**